### DISCUSSION TITLE
[General] How L2 calldata be compressed

### DISCUSSION THREAD

**Discussion by nooma-42** [zkSync-Community-Hub/zksync-developers#255] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

how zksync compress L2 calldata and publish to L1 or simply doesn't do so?
their doc mentioned compression of state diff but nothing about L2 calldata
https://github.com/matter-labs/zksync-era/blob/eb4750d36c86d9a108213974bc6543b77614021e/docs/specs/data_availability/pubdata.md

the tool i got seems to only reconstruct state

### What is the specific issue or error you're encountering?

I tried to read the doc but find no answer

### Can you share the error messages or logs you're receiving, if any?

N/A

### Have you made any recent changes to the contract before encountering this issue?

N/A

### Are there any external libraries or contracts that your contract interacts with?

N/A

### Can you provide the relevant portions of your contract code where the issue is occurring?

N/A

### Have you tried to isolate the problem, and if so, what were the results?

N/A

### What steps have you already taken to try to resolve the issue?

I've tried out this tool and don't know how to interpret some output, I've already send out issue in their repo
https://github.com/eqlabs/zksync-state-reconstruct

**Comment by MexicanAce** [comment, marked as answer]
Hi @NOOMA-42 👋 

For the zkSync Era rollup, we don't publish calldata and instead only publish state-diff. For state verification, we use state diffs. See this video/presentation for further understanding and explanation : https://www.youtube.com/watch?v=PU8jletYF1s

Is there any particular reason you're asking for calldata compression?

**Reply by nooma-42** [reply]
Thank you. I just wanted to confirm the data publishment detail.

**Comment by inpaslimabj** [comment]
# The compression of L2 calldata in zkSync is not explicitly mentioned in the documentation, which primarily focuses on state diff compression for data availability. However, it's possible that L2 calldata may also undergo compression techniques for efficient processing and storage on the L1 chain.
# To understand how L2 calldata is handled and potentially compressed in zkSync, further investigation into the zkSync codebase or reaching out to the zkSync development team for clarification would be necessary. Additionally, exploring tools like zksync-state-reconstruct and examining their output can provide insights into the structure and handling of L2 calldata.

**Comment by batesb188** [comment]
Investigating how zksync compresses L2 calldata for publishing to L1. Documentation mentions compression of state diffs, but details on calldata compression are unclear. Utilizing tools like zksync-state-reconstruct to understand the process and potentially contribute to resolving any issues.

**Comment by tashuraschaefer** [comment]
ZkSync compresses L2 calldata by encoding it into a more efficient format before publishing it to L1. This helps reduce transaction size and gas costs while maintaining data integrity and security.

**Comment by msandiferv** [comment]
It seems like there's uncertainty regarding how zkSync handles compression of L2 calldata when publishing to L1. 
Reviewing the available documentation and tools has not provided a clear answer. Further investigation or reaching out to the zkSync team may be necessary for clarification.

**Comment by m4ccmocare** [comment]
The user is trying to understand how zksync compresses L2 calldata and publishes it to L1, but is unable to find a clear answer in the documentation. They have tried using a tool to reconstruct the state, but are unsure how to interpret the output. They have sent an issue to the tool's repository. No repository link was provided.
